[clang] Distinguish unresolved templates in UnresolvedLookupExpr

zyn0217 opened the pull request:
This patch revolves around the misuse of UnresolvedLookupExpr in
BuildTemplateIdExpr.
    
Basically, we build up an UnresolvedLookupExpr not only for function
overloads but for "unresolved" templates wherever we need an expression
for template decls. For example, a dependent VarTemplateDecl can be
wrapped with such an expression before template instantiation. (See
https://github.com/llvm/llvm-project/commit/617007240cbfb97c8ccf6d61b0c4ca0bb62d43c9)
    
Also, one important thing is that UnresolvedLookupExpr uses a "canonical"
QualType to describe the containing unresolved decls: a DependentTy is
for dependent expressions and an OverloadTy otherwise. Therefore, this
modeling for non-dependent templates leaves a problem in that the expression
is marked and perceived as if describing overload functions. The consumer then
expects functions for every such expression, although the fact is the reverse.
Hence, we run into crashes.
    
As to the patch, I added a new canonical type "UnresolvedTemplateTy" to
model these cases. Given that we have been using this model (intentionally or
accidentally) and it is pretty baked in throughout the code, I think
extending the role of UnresolvedLookupExpr is reasonable. Further, I added
some diagnostics for the direct occurrence of these expressions, which
are supposed to be ill-formed.

As a bonus, this patch also fixes some typos in the diagnostics and creates
RecoveryExprs rather than nothing in the hope of a better error-recovery
for clangd.
    
Fixes https://github.com/llvm/llvm-project/issues/88832
Fixes https://github.com/llvm/llvm-project/issues/63243
Fixes https://github.com/llvm/llvm-project/issues/48673

zyn0217 committed:
[clang] Distinguish unresolved templates in UnresolvedLookupExpr

This patch revolves around the misuse of UnresolvedLookupExpr in
BuildTemplateIdExpr.

Basically, we build up an UnresolvedLookupExpr not only for function
overloads but for "unresolved" templates wherever we need an expression
for template decls. For example, a dependent VarTemplateDecl can be
wrapped with such an expression before template instantiation. (See
https://github.com/llvm/llvm-project/commit/617007240cbfb97c8ccf6d61b0c4ca0bb62d43c9)

Also, one important thing is that UnresolvedLookupExpr uses a "canonical"
QualType to describe the containing unresolved decls: a DependentTy is
for dependent expressions and an OverloadTy otherwise. Therefore, this
modeling for non-dependent templates leaves a problem in that the expression
is marked and perceived as if describing overload functions. The consumer then
expects functions for every such expression, although the fact is the reverse.
Hence, we run into crashes.

As to the patch, I added a new canonical type "UnresolvedTemplateTy" to
model these cases. Given that we have been using this model (intentionally or
accidentally) and it is pretty baked in throughout the code, I think
extending the role of UnresolvedLookupExpr is reasonable. Further, I added
some diagnostics for the direct occurrence of these expressions, which
are supposed to be ill-formed.

As a bonus, this patch also fixes some typos in the diagnostics and creates
RecoveryExprs rather than nothing in the hope of a better error-recovery
for clangd.

Fixes https://github.com/llvm/llvm-project/issues/88832
Fixes https://github.com/llvm/llvm-project/issues/63243
Fixes https://github.com/llvm/llvm-project/issues/48673

zyn0217 commented on clang/test/SemaTemplate/template-id-expr.cpp:
```diff
@@ -186,3 +186,74 @@ class E {
+  // expected-note@#add_pointer {{'add_pointer' declared here}}
+  // expected-error@#ill-formed-decl {{expected ';' after expression}}
+  // expected-error@#ill-formed-decl {{no type named 'type' in the global namespace}}
+  // expected-error@#ill-formed-decl {{'std::add_pointer' is expected to be a non-type template, but instantiated to a class template}}
```
RFC: this diagnostic is probably not optimal (and maybe confusing), since `CheckPlaceholderExpr` is called almost everywhere apart from a full expression. I was considering moving the diagnostic out of that function, but that requires much work I think. (and I'm a bit lazy :)

erichkeane replied:
This is a case where just doing the `-N` positioning is fine, since it is on the same thing here.  

As far as this diagnostic, can you spend a little time to evaluate how much work it is to fix that?  It is particularly awkward.

erichkeane replied:
I think that would be OK?  But  @cor3ntin would be our expert there.

zyn0217 commented:
The pre-commit CI is still clogging after a few hours, and I'm opening it for feedback anyway.

Sirraide reviewed (commented):
I think the general approach of this makes sense. 

Can you add some more tests that involve an expression with this new type as the argument to a function call? You’ve taken care of handling that case from what I’ve seen, but having tests for that would probably be a good idea.

zyn0217 committed:
Add tests for function arguments as well as a typo fix

zyn0217 committed:
cleanup

zyn0217 commented on clang/test/SemaTemplate/template-id-expr.cpp:
```diff
@@ -186,3 +186,74 @@ class E {
+struct Q {
+  template <typename T> int foo() {
+    return T::template I<int>; // expected-error {{'P::I' is expected to be a non-type template, but instantiated to a type alias template}}
+    // expected-note@#TypeAlias {{type alias template declared here}}
```
I think this is more of a preference - subjectively, I feel like using markers makes the diagnostics concentrated, and one can see them in a somewhat consistent block without guessing what else notes/warnings an error diagnostic would also provide, though this looks a bit untidy. :(

zyn0217 replied:
> Can you add some more tests that involve an expression with this new type as the argument to a function call?

Done. Thanks for the feedback.

erichkeane replied:
I VASTLY prefer the markers.  It makes it easier to determine which error is causing the note.

Sirraide replied:
>  It makes it easier to determine which error is causing the note.

Iirc last time this came up in a discussion the consensus was that it was better not to use markers, but I think that was more about actual errors than notes, because this makes sense, yeah.

erichkeane replied:
Yeah, it doesn't make sense for errors/warnings: Those should be "right next" to the thing that caused them (so I prefer the  +/- Numeral for those).  However, with Notes, it makes a TON of sense, and IMO, should nearly ALWAYS be used.

Without them, we end up getting 'errors' and their notes are completely unintelligible, because they are spread out around the file and in no reasonable order.  It makes it so figuring out "this is an error only because of these instantiations" is impossible.

Sirraide replied:
That’s true; I’ve run into cases before where the notes just ended up being a mess...

erichkeane replied:
FWIW, I think I'm the 'first' person who has really been pushing for using the bookmarks for notes, no one ever has before, but when reviewing template stuff in particular, it is REALLY bad, so as code-owner, its one of the things I push for in new commits :)  

So you'll see very few that do it, but hopefully more happening.

zyn0217 committed:
Handle TRANSFORM_TYPE_TRAIT_DEFs in ParseOptionalCXXScopeSpecifier

... so that these "keywords" can be parsed as ordinary identifiers
in a specifier.

zyn0217 committed:
cleanup

zyn0217 commented:
@cor3ntin @erichkeane Any chance we can move forward with this review? I appreciate your feedback!

erichkeane reviewed (commented):
This is introducing a new type that has a pretty subtle difference vs others.  Can we have some documentation in the internals manual explaining the difference?

Code wise, I think this looks fine.

zyn0217 committed:
Add comments to UnresolvedLookupExpr

zyn0217 commented:
> This is introducing a new type that has a pretty subtle difference vs others. Can we have some documentation in the internals manual explaining the difference?
> 
> Code wise, I think this looks fine.

Yeah, I added some explanation to the document of `UnresolvedLookupExpr`; further, I didn't see anything specific to the topic of `UnresolvedLookupExpr` in the `InternalsManual.rst`, so I hope this is sufficient?

zyn0217 committed:
Use '\code' ... '\endcode' block

zyn0217 committed:
Merge branch 'main' into PR88832

zyn0217 commented:
@aganea Oops, this is an oversight I think, because I don't build lldb locally nor run its tests...
I'm proposing a PR https://github.com/llvm/llvm-project/pull/91132, can you help me confirm if that works?

jyu2-git commented:
Hi @zyn0217.
It seems this change cause https://godbolt.org/z/311nb6xYe to fail.  Could you please take look?

Thanks.

zyn0217 commented:
@jyu2-git Are you sure this is the patch you’re looking for? The diagnostic there is completely unrelated and I think the right patch is https://github.com/llvm/llvm-project/pull/90152.

jyu2-git commented:
Hi @zyn0217, I am not sure.  But if I revert your patch, two tests compiles.

zyn0217 commented:
> It seems this change cause https://godbolt.org/z/311nb6xYe to fail. Could you please take look?

> Also:
> https://godbolt.org/z/bYs7Y9v11

I don't see any justification that these examples should compile: the first example is a typical one handled by #90152
(from the release note):
```rst
- Clang now looks up members of the current instantiation in the template definition context
  if the current instantiation has no dependent base classes.

     template<typename T>
     struct A {
       int f() {
         return this->x; // error: no member named 'x' in 'A<T>'
       }
     };

```
and the second one would also be rejected by GCC once the inner class `b` gets instantiated. This earlier diagnostic is probably also triggered by #90152, but I'm not sure. @sdkrystian, is this behavior intended and from your recent patches?
